### ISSUE TITLE
Show Detailed Infos on CI linke circleCI

### ISSUE THREAD

**Issue by muescha** [prettier/prettier#6885]
more people submit PR for documents in MD files or other things via Github web editor. if they make a mistake then the CI fails but there is no change to inform the user whats wrong. 

i see some infos should be with the `--debug-check` but currently there is no output with this option.

**Environments:**
- Prettier Version: 1.18.2
- Running Prettier via: cli
- Runtime: node?
- Operating System: linux

**Steps to reproduce:**

prettier "**/*.{md,css,scss,yaml,yml,ts}" "--check"

**Expected behavior:**

show each error and the changes like with --debug-check here in #4415 

**Actual behavior:**

show nothing about the errors:

```
> @ prettier /home/circleci/project
> prettier "**/*.{md,css,scss,yaml,yml,ts}" "--check"

Checking formatting...
docs/contributing/environment.md
Code style issues found in the above file(s). Forgot to run Prettier?
```

also this are no infos:

prettier "**/*.{md,css,scss,yaml,yml,ts}" "--debug-check"
prettier "**/*.{md,css,scss,yaml,yml,ts}" "--debug-check" "--check"

**Comment by dylannz-sailthru** [comment]
"Forgot to run Prettier?" is really not a very helpful message...

**Comment by thorn0** [comment]
The message contains the file names. What else do you need to run Prettier on those files? ¯\_(ツ)_/¯

**Comment by lydell** [comment]
I personally like the output of `--check` very much.

_CI:_ "Forgot to run Prettier?"

_Me:_ "Oh yeah, I guess I did!" _Runs Prettier._

_CI:_ "All good!"

**Comment by muescha** [comment]
I like to point again to:

> more people submit PR for documents in MD files or other things via Github web editor.

That means they have no local IDE installed. That's why they *can not* run prettier on the file.

**Comment by muescha** [comment]
Expected behavior would be a detailed error message or a diff 

```
[error] src/index.js: ast(input) !== ast(prettier(input))
[error] Index:
[error] ===================================================================
[error] ---
[error] +++
[error] @@ -17,6 +17,6 @@
[error]                    "method": false,
[error]                    "key": {
[error] -                    "type": "StringLiteral",
[error] -                    "value": "asd"
[error] +                    "type": "Identifier",
[error] +                    "name": "asd"
[error]                    },
[error]                    "computed": false,
[error]
[error] Index:
[error] ===================================================================
[error] ---
[error] +++
[error] @@ -1,1 +1,1 @@
[error] -const test = { "asd": 1 };
[error] +const test = { asd: 1 };
[error]
```

**Comment by evilebottnawi** [comment]
Why not run `prettier \"{**/*,*}.{js,json,md,yml,css}\" --list-different"`?

**Comment by muescha** [comment]
thx. I would try on monday the output of `--list-different`

**Comment by lydell** [comment]
`--list-different` also just lists file names – I don’t think it’ll make any difference for @muescha.

@muescha I think this proposal might be solving your problem the wrong way. Forcing people to click into a CI build log, find diffs of expected output and then manually applying them does no sound very friendly. I think we could try several other things:

- Encourage people to clone the project rather than using the web editor.
- Use `--prose-wrap preserve` or `--prose-wrap never` to reduce chance of needing to run Prettier.
- Running Prettier in the web editor via https://github.com/prettier/prettier-chrome-extension.
- A bot that automatically runs Prettier on PRs.
- You could pull people’s branches, run Prettier yourself and push to their PR (contributions from maintainers is generally enabled).
- Not use Prettier for markdown files if you get a lot of contributions to them via the web editor.

**Comment by muescha** [comment]
> - Encourage people to clone the project rather than using the web editor.

thats much overhead :(

> - Use --prose-wrap preserve or --prose-wrap never to reduce chance of needing to run Prettier.

it should run normal prettier

> - Running Prettier in the web editor via prettier/prettier-chrome-extension.

this is only the solution for one user, and not for all 

> - A bot that automatically runs Prettier on PRs.

this changes the PR - the user should have information about the problems

> - You could pull people’s branches, run Prettier yourself and push to their PR (contributions from maintainers is generally enabled).

to much overhead to checkout a branch etc

> - Not use Prettier for markdown files if you get a lot of contributions to them via the web editor.

also prettier is a good help


the only thing is that there should be an more verbose error message :/

**Comment by thorn0** [comment]
@muescha How do you imagine those "verbose error messages"? What exactly could Prettier write in them?

**Comment by lydell** [comment]
Thanks for the responses!

You didn’t seem to respond to this one, though:

> Forcing people to click into a CI build log, find diffs of expected output and then manually applying them does no sound very friendly.

Thoughts?

**Comment by muescha** [comment]
since they edit the PR on github and check why the CI build is failing they already looking into the CI logs. and then when they only find a not helpful message - that is frustrating.

example https://github.com/gatsbyjs/gatsby/pull/18719#discussion_r336781550

> I checked what you suggested and lines where only word to highlight is "false" (without quotes or 0 are highlighted with no issues.
> 
> PS. @muescha is there a way to get rid of prettier complains in build errors? I am editing this in the browser.

**Comment by dylannz-sailthru** [comment]
I was making changes to a repository that'd had prettier added to it without me realising, but I had not installed it in my local development environment. So naturally it blew up in CI and asked if I'd "remembered to run prettier". But that message came from a program called "prettier", which was pretty confusing - yes I suppose I'd "forgotten" to run prettier (I hadn't actually, I didn't even know I had to. Unnecessarily snarky error message, if you ask me...) but even more annoying was that the *prettier* program was printing that message!

`exit 1` would have been just as helpful, and less condescending. A list of formatting violations would have been a lot better though!

**Comment by thorn0** [comment]
@dylannz-sailthru Would you really prefer to get [a diff instead like @muescha has proposed](https://github.com/prettier/prettier/issues/6885#issuecomment-554256895) and apply it manually instead of running the formatter?

**Comment by dylannz-sailthru** [comment]
> @dylannz-sailthru Would you really prefer to get [a diff instead like @muescha has proposed](https://github.com/prettier/prettier/issues/6885#issuecomment-554256895) and apply it manually instead of running the formatter?

I admit, I've now set up my IDE with prettier and now it doesn't bother me. But it seems there is a use case where someone wants to make a small change and doesn't have prettier installed (either on GitHub or in a local development environment), adds an extra space somewhere, and now has to install a 3rd party program and run it instead of just fixing a small error. And while I'll admit it's a narrow use case, I can't see much of a downside to being more verbose (or at least having a configurable verbose flag). Most CI tools will collapse each CI step so you can view the logs of each one in isolation.

**Comment by muescha** [comment]
example of a CI log message:

https://circleci.com/gh/gatsbyjs/gatsby/283065?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

**Comment by thorn0** [comment]
@dylannz-sailthru Locally, extra setup isn't needed. Prettier is installed as a dev dependency and there is usually a command in the project to run all the linters, formatters, etc. (e.g. `npm run format`). This means we're discussing only the web UI use case here.

**Comment by lydell** [comment]
> Unnecessarily snarky error message, if you ask me...

Sorry, the message is actually not supposed to be snarky, but friendly! Do you have ideas for better wording?

> example gatsbyjs/gatsby#18719 (comment)

I don’t think that’s a good example. That’s not a markdown change, but a code change! Is that really something people do in the browser? That sounds terrible to me – how do you debug and test then?

---

Here’s yet one idea: When editing a markdown file in the GitHub editor and are about to hit the Commit button – first copy-paste the entire file into https://prettier.io/playground/ and then copy back the output.

**Comment by muescha** [comment]
> I don’t think that’s a good example. That’s not a markdown change, but a code change! Is that really something people do in the browser?

yes - people do it. so why not also on all files.

> Here’s yet one idea

that maybe a workaround but not a solution and also not fit with the project settings for prettier

**Comment by dylannz-sailthru** [comment]
> Sorry, the message is actually not supposed to be snarky, but friendly! Do you have ideas for better wording?

Sure! Something like: ```You need to run `prettier fix` on these files to ensure they are formatted correctly.```

This way it's telling me more specifically what I need to do to get past the error. I think it doesn't matter so much for people who have used `prettier` before, but as someone who hadn't used it before I found it a little confusing.

> Here’s yet one idea: When editing a markdown file in the GitHub editor and are about to hit the Commit button – first copy-paste the entire file into https://prettier.io/playground/ and then copy back the output.

Actually that'd be a great thing to link to in the error message as well - I'd have found it really useful at the time :)

**Comment by asfaltboy** [comment]
How come html file diffs are outputted, but not JS/CSS files? Is it a verbosity concern? If so, perhaps we should be able to toggle via `--log-level` (i.e set the default to warn and don't display diff on warn)?

**Comment by grncdr** [comment]
I would also like to see a diff or more detailed information about why prettier thinks a file isn't formatted. I currently don't run `prettier --check` in CI/pre-commit hooks because it telss me that files I've just formatted with Prettier are still not formatted. I really don't know how to begin fixing the problem when I can't see what it is.

**Comment by lydell** [comment]
@grncdr It tells you which files aren’t formatted, so all you need to do is run Prettier on those files. Or on your whole project – probably `prettier --write .`. Can you expand on how seeing a diff helps?

**Comment by grncdr** [comment]
As I said in my first message. I have already run prettier on those files, but it still shows the warning. Trying to make a more useful error report is impossible when I can't see what formatting prettier is complaining about.

**Comment by lydell** [comment]
1. Check that you are using the same Prettier version locally and in CI.
2. Check that you are using the same config locally and CI.
3. There’s probably no step 3.

**Comment by grncdr** [comment]
so here's what I just did, all locally, no CI involved. 

```
$ yarn prettier --check some-file.ts
yarn run v1.22.4
$ prettier --check some-file.ts
Checking formatting...
[warn] path/to/some/file.ts
[warn] Code style issues found in the above file(s). Forgot to run Prettier?
error Command failed with exit code 1.
$ yarn prettier --write some-file.ts
yarn run v1.22.4
$ prettier --write some-file.ts
some-file.ts 188ms
✨  Done in 0.48s.
$ git diff
<actual diff output>
```

It turns the diff was caused by another tool formatting & sorting imports, and it disagrees with Prettier about how many imports belong on a single line. This was trivial to see _once I had a diff_ but difficult to understand without it. The experience in VSCode is that format-on-save is using Prettier and everything looks fine, so there's no indication about the other tool messing with Prettiers formatting.

Having an option for more verbose output would have made the problem clear immediately.

**Comment by guyzmo** [comment]
Hello 👋 

I'm having a project which has prettier configured. I run prettier on save in my editor, and I have got husky configured to ensure commits are well formatted. But when I push to gitlab-ci, it reports that a couple of files have issues, telling me that I forgot to run prettier.

When I try to replicate the CI locally (either using a similar docker configuration, or using gitlab-runner) it works well.

> Check that you are using the same Prettier version locally and in CI.

I believe it is (as it's installed and ran through yarn)

> Check that you are using the same config locally and CI.

I believe it is (as it's ran through the package.json scripts)

> There’s probably no step 3.

It's still not working.

If I could have had the details of the issues, that would help me a lot pinpoint the problem:

* is the CI actually building the commit I think it is supposed to build?
* is there a change in the environment when running remotely that changes the prettier configuration?
* is the prettier version (which is installed by yarn with a yarn.lock) the version it is supposed to install (which should be iso with my local installation)?
* is there an issue with the job's cache?
* or am I just damned?

In the end, I believe it would be nice to have a way to have knowledge of what the issues are by making `--check` more verbose when ran in CI for debugging purposes.

**Comment by thorn0** [comment]
@guyzmo 

> I believe it would be nice to have a way ... for debugging purposes

But you have that way, don't you? I mean the combination of `prettier --write` and `git diff`.

**Comment by thernstig** [comment]
I am adding my two cents here, as my duplicate was closed: https://github.com/prettier/prettier/issues/10294

When I have white-space issues in a JavaScript file I get the following error when running via CLI:

```sh
> prettier --check "**/*.{js}"
Checking formatting...
[warn] api/cow/moo.js
[warn] Code style issues found in the above file(s). Forgot to run Prettier?
```

If I run Prettier via ESlint (using eslint-plugin-prettier) I instead get the following output:

```sh
> eslint --max-warnings=0 .

/home/user/code/project/api/cow/moo.js
  9:1  error  Delete `⏎··⏎`  prettier/prettier

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.
```


I find the ESLint output _much_ better than that of Prettier. Hence why I have chosen to run Prettier via ESLint instead. This is unfortunate, as https://prettier.io/docs/en/integrating-with-linters.html does not recommend to use Prettier as plugin to ESlint.

The actual behavior is subpar to what I would expect. I want to know the file, the line:column and also exactly what was wrong.

I'd also add that Prettier is very concerned about formatting/style/"beauty" - yet the CLI output used in e.g. CI is nothing of those. I feel this is important to raise the quality-feel of Prettier.

**Comment by larskinn** [comment]
I'll just +1 the usefulness of a diff for CI.

I'm trying to debug an issue where `prettier --check` lists ALL files as incorrectly formatted when running in CI, while running `prettier --check` locally -- yes, with the same .prettierrc -- shows "All matched files use Prettier code style!"

I'm guessing it's a line ending issue, but it's very hard to tell what's going on when there's no diff in the output. I'm basically just trying various changes to Git's line ending handling and seeing if they make the CI run successfully (no luck so far).

**Comment by thibaudcolas** [comment]
This may be useful to some – you can work around Prettier not providing this by using your git (or other SCM) diffing tools, and running Prettier twice. The first command applies formatting changes, then shows the diff, then discards all the changes. The second command checks the formatting as you normally would.

```sh
prettier --write '**/?(.)*.{md,css,scss,js,json,yaml,yml}' && git --no-pager diff && git checkout -- .
prettier --check '**/?(.)*.{md,css,scss,js,json,yaml,yml}'
```

For people who want the best possible build performance, I’m sure it would also be relatively straightforward to only run Prettier the one time with `--write`, and then break the build if there are local changes in git. Personally I prefer to still preserve the output of `--check`.

---

If Prettier does end up with supporting this, it’d be great to do it in a way that can be integrated with GitHub Actions’ annotations (e.g. for [ESLint](https://github.com/actions/setup-node/blob/c6fd00ceb9747fb23ffdf72987450a2664414867/.github/eslint-stylish.json)).

**Comment by JoshuaKGoldberg** [comment]
I just came across this with a project with weirdly formatted endlines across Windows vs. *Nix. CI failed on a few files that passed Prettier locally. This would be useful.

Additionally, +1 to the comments mentioned about folks who don't have Prettier set up locally. 

@thorn0 what part of this issue would the maintainer group consider to still be in discussion? I'd be keen on implementing this.

**Comment by Trim21** [comment]
you can use [reviewdog](https://github.com/reviewdog/reviewdog)

```yaml
      - uses: reviewdog/action-setup@v1
        with:
          reviewdog_version: latest # Optional. [latest,nightly,v.X.Y.Z]

      - name: lint
        run: |
          prettier --write --list-different ./
          TMPFILE=$(mktemp)
          git diff >"${TMPFILE}"
          git stash -u && git stash drop
          reviewdog -f=diff -f.diff.strip=1 -reporter=github-pr-review < "${TMPFILE}"
        env:
          REVIEWDOG_GITHUB_API_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```

**Comment by theahura** [comment]
Adding my voice to the chorus: my team has been stuck for a few days now because `prettier --check` is failing on Github Actions for files that we explicitly run `prettier --write` on. Locally prettier says there are no files that need changing; on Github Actions, prettier throws errors. 

A debug would be really important to figure out why.

**Comment by lydell** [comment]
@theahura How would a diff help you solve that problem (local and GitHub Actions disagree)? And what do you mean by “prettier throws errors”?

**Comment by theahura** [comment]
By 'prettier throws errors' I mean that prettier --check indicates that there are problems.

A diff would provide a starting point to find patterns for changes. For example, I used the git diff example above and noticed that the changes were all due to spacing. I was then able to Google search to find that there was a prettier version change that was causing issues. Without the git diff, I basically had no launching off point to figure out what the issue was. I might have eventually landed on the version change, but only after checking a lot of other things.  

Prettier is great, and I understand the desire to keep the UI simple. But I imagine you're already calculating what changes need to be made in order to check and write them; is it hard to dump those to stdout, even in a messy format?

**Comment by lydell** [comment]
> I was then able to Google search to find that there was a prettier version change that was causing issues.

Checking the version of Prettier locally and CI would have been the first thing I had checked. That’s usually the cause for all types of CI issues I encounter (Prettier related or not) 😅

**Comment by theahura** [comment]
Maybe the output of the prettier check should include a note that says 'if you're still seeing errors, make sure your version is correct'?

But either way I think the goal should ideally be to make people self sufficient so they don't have to ask you. One way to do that is to hide power user options behind flags like --debug or --verbose or whatever. In my case I went to check versions only after exhausting a bunch of other possibilities (I spent a long time fiddling with the prettierrc, for eg). The diff helped me realize that it wasn't the rc.

**Comment by webark** [comment]
An example of how I found myself on this issue, but in the end, it doesn't matter.

I came in new to a project, and it was giving me an annoying little red squiggle when I was writing something a certain way. I was like, "Ohh, some has some annoying rule is on.. Uhhgg. People. 🙄" (sorry.. I'm opinionated about my code style (something you authors should appreciate 😉 )). So I went to see what the rule was, cause naturally, I wanted to disable it, and all I got was "prettier/prettier". After another (opinionated) eye roll, I was like "ohh I'll just run the command and it will output the rule that I can go disable!"

So I ran the command and got the "friendly" `Forgot to run Prettier?` output. (Now a quick aside, assuming that someone "forgot" to do something can be offputting if you purposefully did not do what they are accusing you of "forgetting". No, I did not "forget" to run prettier, I "chose" not to run prettier.) Assuming I was running the "light" output, I went to look to see all the different options and discovered, no, this is the most detailed output.

I then thought to myself, "Ohh, maybe I can just look over the options and find which one!". After skimming over the options, being surprised that was all there was, assuming the documentation was incomplete I started googling around. And that's when, after finding my way back to the config page, saw the "Option Philosophy" link, and then I read the _dreaded_

> **But we won’t add more of them.**

This is when I realized that ohh, I'm just going to have to live with this (again in my opinion) substandard syntax. 😞 🤡 

> Sorry, the message is actually not supposed to be snarky, but friendly! Do you have ideas for better wording?

@lydell I opened up PR #11369 with an updated message. I wanted to keep it the same succinctness, just being a little blander and informative, and a little less familiar so it can't be "misinterpreted" as easily. Any additional information (such as a  doc link or an example command) seemed like a separate initiative that could happen regardless if this change was made.

Thanks for the work, I know OSS isn't a glamour gig. 😊

**Comment by worc** [comment]
the diff is also important when you're verifying that prettier isn't causing breaking changes or just plain bad legibility. it's also important to discovering all the unwritten rules. i'm not here to blindly apply "fixes", i want to know what changes prettier is proposing so i can incorporate that into my work on repos that use prettier in the future.

a good example is the strong opinion prettier has on css/sass grid styles: https://github.com/prettier/prettier/issues/2703

i'd rather be shown what prettier wants to do and decide if that's the right course of action, instead of the condescending `Forgot to run Prettier?`

**Comment by liamjsc** [comment]
Is it possible for the CLI output to list the prettier config rules that are broken? I'm setting up my prettier config and this snippet changes from

```
get: param => ({ anotherParam }) => {
    const foo = bar;
```

to this:

```
  get:
    param =>
    ({ anotherParam }) => {
      const foo = bar;
```

I much prefer the single-line params, and I'd like to know which rule I can set to disable splitting this into multiple lines, but the only output I get from prettier-check is "Code style issues found in the above file(s). Forgot to run Prettier?"

**Comment by lydell** [comment]
@liamjsc You can:

1. Commit all your code unformatted.
2. Run Prettier with `--write`.
3. Check the (git) diff.
4. Discard all changes.
5. Make changes to the config.
6. Repeat from 2.

**Comment by liamjsc** [comment]
Thanks for the reply @lydell , but what changes to the config are you suggesting? Maybe my question was not clear enough.
--write is not telling me which prettier rules are broken.

For example, if prettier was changing my code from:
`const foo = bar` to `const foo = bar;`, how can I discover that the rule being applied is `semi: true`?

**Comment by lydell** [comment]
I’m not suggesting any config changes, I just tried to help with a workflow where you can see the diff of what Prettier does.

There aren’t that many options – going through https://prettier.io/docs/en/options.html should make it fairly obvious what each might do. The [playground](https://prettier.io/playground/) can help, too, with playing around with different options.

**Comment by Israel-Laguan** [comment]
I had an issue similar to this one.

So I added a file `.gitattributes` like this:

```sh
*.ts text eol=lf
```

With this the problem was solved, but I think prettier should have a more descriptive message

**Comment by worc** [comment]
Another use-case is when doing a rebase and re-linting / re-formatting a conflict resolution. I'd rather see what needs changing instead of blindly applying fixes in the midst of a rebase.

**Comment by masalomon** [comment]
My suggestion would be to keep the current behavior for `--check`, but add another option to check and display a diff, or perhaps add a `--verbose` option for that purpose (i.e. `--check --verbose`).  Alternatively, the potential workaround suggested by [this comment](https://github.com/prettier/prettier/issues/6885#issuecomment-1045233499), above, could be added to the documentation (along with a description of what running `prettier` with no options does).

For example:
> To highlight the changes to a single file that `--write` would make, you can run
`$ diff filename <(npx prettier filename)`  (using process substitution, on *ix style shells)
Or, to use `git diff`'s features (`git-diff` doesn't deal with process substitution), run
`$ npx prettier [options] filename | git diff [--no-index] [other options] filename -`

**Comment by meza** [comment]
I would very much like to see something like this too. It's one thing to enforce a standard but it's another to educate people. Telling me that I've messed something up without telling me what it is that I've messed up or why it's a bad thing in the first place is not helpful. If someone relies on `prettier --write` 100% of the time, they might never know that they are making obvious mistakes. Now some might argue that they don't even care because prettier will fix it. I can accept that.
However for the rest of us it's just not a useful behaviour. I'd just like to learn and become a better engineer by knowing more about the conventions I work with.

**Comment by salmiakki** [comment]
`black` (a popular formatter for Python) has the `--diff` option to show the changes it would make. This option is super handy, I used it a lot. Not so much lately. Because it has taught me to write properly.

**Comment by lydell** [comment]
[dprint](https://dprint.dev/cli/) has diff too (`dprint check`), and so does [gofmt](https://pkg.go.dev/cmd/gofmt) (`gofmt -d`). So it looks like there’s precedent for such a thing!

**Comment by QuintinWillison** [comment]
Adding my +1 for this feature. I'm sitting here trying to work out why Prettier running on my local macOS hosted succeeds while the same version, [running on a Linux host in CI](https://github.com/ably/features/runs/7832464216?check_suite_focus=true), is failing. The lack of verbose output from Prettier when things go wrong is woeful and, I suspect symptomatic of some high-level, idealistic 'opinionated' vision held by the developers of this tool that simply doesn't match the needs of the larger majority of their user base. Perhaps try listening to your users? Users will only fight the tools for so long before they just go ahead and replace them with something that does meet their needs - I've just not got the time today to do that research, but I might tomorrow. 🤷

**Comment by abdulramonjemil** [comment]
> @grncdr It tells you which files aren’t formatted, so all you need to do is run Prettier on those files. Or on your whole project – probably `prettier --write .`. Can you expand on how seeing a diff helps?

Prettier may do a good job most time, but it still has its own flaws. The other time, I was using Prettier with ESLint, and I used the following in my `.jsx` file.

```jsx
<label htmlFor="some-id">Some label</label> /* eslint-disable-line */
```

The comment after the line is supposed to stop eslint from emmiting an error for that line (as I've used some eslint config that causes an error when the htmlFor prop is used). When formatted with Prettier, the comment is moved to the next line causing ESLint (which shouldn't have thrown an error) to throw an error.

This is why I like to make sure that I see what formatting Prettier is applying to my code. A diff is the solution. If I get to see that, I would adjust the code as needed. And whenever I know that Prettier will do a pretty good job at formatting, I can just run `prettier --write` on the file/directory and if anything goes, I would be the one at fault, not Prettier, as I wasn't forced to do it.

**Comment by lydell** [comment]
@abdulramonjemil I’m not against `--check`, but:

- I recommend running Prettier on save in the editor, and saving often. Then you’d see that eslint-disable comment jump down early.
- Instead of carefully inspecting the diff of `prettier --check` before running `prettier --write`, you could carefully inspect the diff of `git diff` after `prettier --write`.
- ESLint would catch that moved disable comment anyway – because it would start complaining again about whatever you were disabling.

I see a diff in `--check` more of as an UX improvement for CI, not something you’d run all the time. You’re of course free to use `--check` however you like, but I think there are better ways to use Prettier!

(Btw, I bet 80 % of all CI failures are newline related, so a diff implementation in Prettier should take extra care to make newline differences clear.)

**Comment by Osmose** [comment]
@lydell Would a separate `--diff` flag be more acceptable? I'm happy to fix up https://github.com/prettier/prettier/pull/12598 if there's someone to review / merge it.

**Comment by abdulramonjemil** [comment]
That's well said. I just think having the --diff option would be good.

**Comment by muescha** [comment]
how the an option `--diff` should work? It shows the diff only when the check fails?

**Comment by muescha** [comment]
As I see in PR it should used instead of `--check` or can it used in combination: `--check --diff`?

I think it is hard to convince repo owners to change from `check` to `diff` because of the name. Maybe `--check-verbose`  or `--check-and-diff` (some similar name) shows that it is the same check but more info.

**Comment by lydell** [comment]
I don’t get why a new flag is needed. Isn’t it good to always output diff in `--check`? Is it about some kind of backwards compatibility? Again, I’m not against diff in `--check` – I think it’s good UX and worth the extra code to do it, as discovered in this issue.

**Comment by Osmose** [comment]
I thought I read some concern upthread of the compatibility of changing `--check` 's output, but I don't have strong opinions either way; I just want _some_ way of getting diff output that is likely to get reviewed and merged, whatever that may be.

**Comment by abdulramonjemil** [comment]
I agree with you.

**Comment by ChannyClaus** [comment]
can we at least add in the error message that `prettier --write` will make the change that resolves the issue? for someone who is not an experienced user, that will save a lot of time. i just spent about 30 minutes trying to figure out how to fix a failing `prettier --check`.

**Comment by lydell** [comment]
@ChannyClaus You mean something like “Run Prettier with the `--write` flag to fix issues”? That sounds good!

**Comment by webark** [comment]
I opened up this PR which addresses half of it.

I didn't add links or extra example commands due feeling that was more of a separate issue.

It got approved, but never merged. Perhaps doing both at the same time is better.

**Comment by abdulramonjemil** [comment]
Exactly. It's definitely a good thing to be able to see what changes would be made by prettier `--write`.

**Comment by itinance** [comment]
GitLab Ci is failing when prettier has listed warnings. Can we change the way how prettier handles warnings as errors?

What happens on GitLab CI:

```
[warn] Code style issues found in 9 files. Forgot to run Prettier?
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
Cleaning up project directory and file based variables
ERROR: Job failed: exit code 1
```

I'd like to see GitLab CI failing only when real errors occured. Is that somehow possible?

**Comment by webark** [comment]
Prettier doesn't offer a diffing option, nor do i believe they ever show a diff, and there might not be that capability.

I think the work around people have suggested is to run the write and then do a diff, and fail on if a diff is present.

I don't believe a PR has been rejected yet that would offer a diff option to be added to the check command.

**Comment by itinance** [comment]
Strange thing: I could solve it in a very weird workaround:

The actual Issue: running locally in my project-folder/repository, any  `npx prettier --write` did NOTHING. But on GitLab CI the prettier-check failed. 

So I did:

1. I cloned the project-repository on my working machine in a temporary folder. 
2. suddenly `npx prettier -c`  gave the same warnings like in GitLab CI, still NOT in my actual worksapce/project folder
3. I ran `npx prettier --write` and pushed back into the repository
4. GitLab Ci ran through successfully
5. In my actual project folder I did `git pull` in order to get the changes from $3

I'm still curious why `npx prettier -write` did nothing on my project folder and needed a separate checkout.

At least we have now a workaround.

**Comment by daveisfera** [comment]
The `eslint` output in [stylish format](https://eslint.org/docs/latest/user-guide/formatters/#stylish) is useful because many CI tools already parse it and make the information available, so that would be a nice option

**Comment by salmiakki** [comment]
> ...many CI tools already parse it...

@daveisfera Could you name a few?

**Comment by daveisfera** [comment]
> > ...many CI tools already parse it...
> 
> @daveisfera Could you name a few?

GitHub Actions and Bamboo are the two that I know of that can parse this output

**Comment by Jasperrr91** [comment]
@lydell 
> * Instead of carefully inspecting the diff of `prettier --check` before running `prettier --write`, you could carefully inspect the diff of `git diff` after `prettier --write`.

I understand where you are coming from but `prettier --write` is fairly intrusive in that it overwrites all of your files. If users had forgotten to stage/commit files beforehand (and that happens often enough) they cannot simply discard the changes caused by a incorrect configured `prettier`.

Of course, one could edit their prettierconfig and run it again and again till the desired config is reached but still, there's no way to properly revert.

> I see a diff in `--check` more of as an UX improvement for CI, not something you’d run all the time. You’re of course free to use `--check` however you like, but I think there are better ways to use Prettier!

There might indeed be better ways to use Prettier but looking at the amount of comments and votes in this thread, there is clearly a lot of people that would like to use Prettier in this specific way. Now, it's easy to support both ways of working. Simply add a `--verbose` flag so that users that want more detailed logs can run `prettier --check --verbose` while others can stick to the regular `prettier --check`.

I also agree that the current error message is incorrect either way. "Forgot to run Prettier" is a weird error to give to a user that has just ran Prettier.

**Comment by aduh95** [comment]
As a workaround, if your project is using git, you can use `prettier . -c || (git stash push -q; prettier . -w --loglevel silent && git --no-pager diff && git reset --hard; git stash pop -q)` to generate a diff. It would still be worth if Prettier could generate a diff itself of course, but the above command can already improve the DX.

**Comment by shasherazi** [comment]
Still no diffs in outputs???

**Comment by jiri-thiemel-te** [comment]
Listing the rule(s) because of which a file contains warning(s) is necessary to understand the current setup. (E.g. when some behaviour makes you crazy and you want to find out which rule is responsible for that to modify it). I'M SHOCKED that there is no way to list this when running prettier command. How some functionality as basic and expected as this one can be missing is beyond me.

**Comment by abhijeetviswa** [comment]
I agree with others that this is some pretty  basic functionality that is provided by tooling in other languages (isort, black in Python). At the very least, it'd be useful to know as a beginner to the codebase what you messed up when it comes to the style so that you can make a conscious effort at not repeating it instead of relying on the IDE to magically fix things for you. 

For me style checks in a CI env is akin to UTs. Think of how annoying and frustrating it would be if your test suite said "A test failed in your project. Did you not write it properly?" without giving you ANY further logs on which test failed.

**Comment by mukeshsharma1201** [comment]
I encountered this issue today! 
I am a backend dev and needed to do a quick bug fix in the frontend codebase.
I made one line change and pushed the code to find that the CI is failing with the message `Code style issues found in the above file. Forgot to run Prettier?`

I don't want to install whole npm and frontend packages on my machine just to find out 1 small issue that I could have fixed right away if there was a simple diff or description about the issue! 

And this issue is Open for 4 years now!!

**Comment by maxolasersquad** [comment]
Agreed. Just saying "something" failed on the failed, without saying what is not as useful as it should be. For example, eslint gives output like this.

```
/home/foo/projects/src/components/Foo.js
  33:1  error  `FooInput` import should occur before import of `HamburgerInput`  import/order

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.
```

Not only does it tell me which file had a failure, but it also tells me at what line, and what the failure is. If I'm confused, I can search the web for "eslint import/order" for additional information. A similar output from prettier would be very helpful.

My current method of understanding prettier failures is to run with the `--wirite` option, and then looking at the git diff, which is not preferable.

**Comment by ngerritsen** [comment]
So we have our Prettier setup correctly with our developers IDE's, but now we are running into some issues in the CI pipeline. It's probably some cache issues or something, also probably unrelated to Prettier itself.

But anyway, whatever the actual issue is, it would be quite helpful if we could see what the problem in a file is in the CI logs, because on our local IDE's/terminals we don't have the issue. It would make debugging why our CI is failing much faster, we could see if it's maybe running it on an old version of the file or whatever. I think it's a pretty mundane feature that we should have.

**Comment by webark** [comment]
Have you tried the "Prettier write -> git diff" method?

**Comment by maxolasersquad** [comment]
The "Prettier write -> git diff" method doesn't really work in CI pipelines. The container running the pipeline probably doesn't even have git installed.

**Comment by webark** [comment]
I mean.. Depends on the pipelines. There are a lot of pipelines that do some validation, and then generate a git tag and pushing it.

**Comment by george-gca** [comment]
I was finally able to overcome this limitation with GitHub actions using `prettier --write`, `git diff`, and [diff2html](https://github.com/rtfpessoa/diff2html-cli). I created 2 GitHub actions: one to run on PRs and another to run on pushes. I used `git diff -- . ':(exclude)package-lock.json' ':(exclude)package.json'` because installing `prettier` currently changes these 2 files, so it gives a false positive in the diff.

The `on push` tests if the code is passing `prettier --check`. If it isn't it creates a git diff as an html file and upload it as an action artifact. The artifact currently is a zip file with the html diff inside, since viewing artifacts directly in browser is [currently not supported](https://github.com/actions/upload-artifact/issues/14).

```yml
name: Prettier code formatter (Push)

on:
  push:
    branches:
      - master
      - main
  workflow_dispatch:

jobs:
  check:
    # available images: https://github.com/actions/runner-images#available-images
    runs-on: ubuntu-latest
    steps:
      - name: Checkout 🛎️
        uses: actions/checkout@v3
      - name: Setup Node.js ⚙️
        uses: actions/setup-node@v4
      - name: Install Prettier 💾
        run: npm install --save-dev --save-exact prettier @shopify/prettier-plugin-liquid
      - name: Prettier Check 🔎
        id: prettier
        run: npx prettier . --check
      - name: Show diff 📝
        # https://docs.github.com/en/actions/learn-github-actions/expressions#failure
        if: ${{ failure() }}
        run: |
          npx prettier . --write
          git diff -- . ':(exclude)package-lock.json' ':(exclude)package.json' > diff.txt
          npm install -g diff2html-cli
          diff2html -i file -s side -F diff.html -- diff.txt
      - name: Upload html diff
        if: ${{ failure() && steps.prettier.conclusion == 'failure' }}
        uses: actions/upload-artifact@v4
        with:
          name: HTML Diff
          path: diff.html
          retention-days: 3
```

The `on PR` one adds a comment to the PR with the html content of the diff. This one I haven't tested yet since it involves this PR being accepted in the repo I commited it and then a next PR failing prettier test.

```yml
name: Prettier code formatter (PR)

on:
  pull_request:
    branches:
      - master
      - main
  workflow_dispatch:

jobs:
  check:
    # available images: https://github.com/actions/runner-images#available-images
    runs-on: ubuntu-latest
    steps:
      - name: Checkout 🛎️
        uses: actions/checkout@v3
      - name: Setup Node.js ⚙️
        uses: actions/setup-node@v4
      - name: Install Prettier 💾
        run: npm install --save-dev --save-exact prettier @shopify/prettier-plugin-liquid
      - name: Prettier Check 🔎
        id: prettier
        run: npx prettier . --check
      - name: Show diff 📝
        # https://docs.github.com/en/actions/learn-github-actions/expressions#failure
        if: ${{ failure() }}
        run: |
          npx prettier . --write
          git diff -- . ':(exclude)package-lock.json' ':(exclude)package.json' > diff.txt
          npm install -g diff2html-cli
          diff2html -i file -s side -F diff.html -- diff.txt
      - name: PR comment with html diff
        # https://docs.github.com/en/actions/learn-github-actions/expressions#failure-with-conditions
        if: ${{ failure() && steps.prettier.conclusion == 'failure' }}
        uses: thollander/actions-comment-pull-request@v2
        with:
          filePath: diff.html
```

**Comment by automaton82** [comment]
I'm a back-end dev coming mostly from C# and am fairly new to npm, node etc. I was surprised to see a linter not give information on what the issues are in a check, mostly because I cannot blindly allow it to modify files.

I have used eslint previously and definitely like the output from it very much. It's short, concise and tells you exactly what to do to fix it if you want to do it yourself.

Glad to see this ticket open (since 2019🫠) and yes a `diff` or really anything would help at this point.

**Comment by muescha** [comment]
also there exists an Feature Request for `reporter`:

- Adding support for reporters #11833

**Comment by george-gca** [comment]
Yes, but it does not solve this issue, as [pointed](https://github.com/prettier/prettier/pull/12598#issuecomment-1090932634) by @Osmose.

**Comment by thatsIch** [comment]
I found, that sometimes there are false positives or new situations where you do not want the code formatted like that. There is no way, that `--write` is 100% correct all the time. There is no feedback loop if you want to learn the rules.

https://github.com/prettier/prettier/pull/12598 would solve the conflict.

It does not modify the default "way" of prettier.

**Comment by siawyoung** [comment]
> Struggling with an issue where locally running `prettier --check` yields no errors and `prettier -w` yields no diff, yet in GitHub Actions, using Ubuntu 22, `prettier --check` fails on a single `tsconfig.json`.
> 
> I think this is definitely an example of a legit use case for getting a diff printout straight from Prettier. Yes, it can be achieved with a combination of `prettier -w` and `git diff` but that's assuming Prettier is used in a Git repo which might not always be the case.

Oh man, this is happening to me now. It only happens for tsconfig.json files too! Not sure why.

**Comment by webark** [comment]
@alexey-sh BLASPHEMER!! 😂🙃

Another option would be to use the eslint plugin, which I think will show the errors when you run it through that?

**Comment by jballands** [comment]
> Another option would be to use the eslint plugin, which I think will show the errors when you run it through that?

I want to note that some folks feel that [integrating Prettier into your linter is suboptimal](https://prettier.io/docs/en/integrating-with-linters):

> - You end up with a lot of red squiggly lines in your editor, which gets annoying. Prettier is supposed to make you forget about formatting – and not be in your face about it!
> - They are slower than running Prettier directly.
> - They’re yet one layer of indirection where things may break.

IMHO, whatever works best for your workflow is what you should do. I don't know if there's a perfect answer here, but using the [eslint-plugin-prettier](https://github.com/prettier/eslint-plugin-prettier), as suggested above, will get you what you're looking for. I would not expect the Prettier team to integrate this functionality; it feels quite intentional.

**Comment by vjekoart** [comment]
Is there at least an explanation why the Prettier team doesn't want to add `--dry-run`, `--diff` or any other similar option to the CLI?

I understand that Prettier is built with certain philosophy in mind, but there were some pretty valid points in this thread why one would benefit from verbose explanation of style errors.

**Comment by lvjp** [comment]
Hello @lydell,

It appears that @muescha has stopped the fight...

In your repl

> Forcing people to click into a CI build log, find diffs of expected output and then manually applying them does no sound very friendly.

**Comment by lvjp** [comment]
Hello @lydell,

It appears that @muescha has stopped the fight...

In one of your [reply](https://github.com/prettier/prettier/issues/6885#issuecomment-554628190) you said :

> Forcing people to click into a CI build log, find diffs of expected output and then manually applying them does no sound very friendly.

There are plenty of people who work in this way, what's the matter?

But this is not the only one use case.  

When you use a linter collection app like the GitHub super-linter.  
If the CI end's up with an error and asks you to use the `--write` argument.
You can either refactor your CI pipeline to add the `--write` argument and `git diff` command to show the diff, or force the developer to install it on the local environment.

What if you would rather not install it in your environment for security reason or simply to not handle updates...

What if you check not yet committed code, you may not want to have some changes done from a linter.

**Comment by lydell** [comment]
@lvjp I wrote that 5 years ago. Since then I’ve changed my mind:

https://github.com/prettier/prettier/issues/6885#issuecomment-1202964469
https://github.com/prettier/prettier/issues/6885#issuecomment-1257048736
https://github.com/prettier/prettier/issues/6885#issuecomment-1257067145

(You might also have noticed that I’m not actively involved with this project anymore.)

**Comment by lvjp** [comment]
Please forgive me for bothering you with problems from the past, and thank you for your swift response.

Maybe @fisker has an opinion on the subject?

**Comment by muescha** [comment]
> It appears that @muescha has stopped the fight...

@lvjp I don't give up, but repeating my arguments every month isn't helpful ;)

**Comment by Osmose** [comment]
I remain ready to rebase and fix up the PR for this once someone with merge access commits to reviewing it 🫡

**Comment by lvjp** [comment]
> I remain ready to rebase and fix up the PR for this once someone with merge access commits to reviewing it 🫡

@Osmose > can you add the link to the PR here ?

**Comment by Osmose** [comment]
> @Osmose > can you add the link to the PR here ?

It's been linked before above, it's https://github.com/prettier/prettier/pull/12598

**Comment by mark-wiemer** [comment]
Bumping @george-gca's solution from January 2024 as it works great for me, seems to be the best workaround available, and didn't get any attention. Easily adaptable to other projects as well. In my case, the issue was the git was replacing LF with CRLF. I should've figured as much, but without this tool it was a blind search (git wasn't warning me on my local machine).

Ref sample run: https://github.com/mark-wiemer-org/vscode-autohotkey2-lsp/actions/runs/11063073280/job/30738602614?pr=29.

I've also adapted the code to my repo and removed some emojis and shopify stuff for simplicity and better practice: https://github.com/mark-wiemer-org/vscode-autohotkey2-lsp/actions/runs/11063073280/workflow?pr=29

> I was finally able to overcome this limitation with GitHub actions using `prettier --write`, `git diff`, and [diff2html](https://github.com/rtfpessoa/diff2html-cli). I created 2 GitHub actions: one to run on PRs and another to run on pushes. I used `git diff -- . ':(exclude)package-lock.json' ':(exclude)package.json'` because installing `prettier` currently changes these 2 files, so it gives a false positive in the diff.
> 
> The `on push` tests if the code is passing `prettier --check`. If it isn't it creates a git diff as an html file and upload it as an action artifact. The artifact currently is a zip file with the html diff inside, since viewing artifacts directly in browser is [currently not supported](https://github.com/actions/upload-artifact/issues/14).
> 
> ```yaml
> name: Prettier code formatter (Push)
> 
> on:
>   push:
>     branches:
>       - master
>       - main
>   workflow_dispatch:
> 
> jobs:
>   check:
>     # available images: https://github.com/actions/runner-images#available-images
>     runs-on: ubuntu-latest
>     steps:
>       - name: Checkout 🛎️
>         uses: actions/checkout@v4
>       - name: Setup Node.js ⚙️
>         uses: actions/setup-node@v4
>       - name: Install Prettier 💾
>         run: npm install --save-dev --save-exact prettier @shopify/prettier-plugin-liquid
>       - name: Prettier Check 🔎
>         id: prettier
>         run: npx prettier . --check
>       - name: Show diff 📝
>         # https://docs.github.com/en/actions/learn-github-actions/expressions#failure
>         if: ${{ failure() }}
>         run: |
>           npx prettier . --write
>           git diff -- . ':(exclude)package-lock.json' ':(exclude)package.json' > diff.txt
>           npm install -g diff2html-cli
>           diff2html -i file -s side -F diff.html -- diff.txt
>       - name: Upload html diff
>         if: ${{ failure() && steps.prettier.conclusion == 'failure' }}
>         uses: actions/upload-artifact@v4
>         with:
>           name: HTML Diff
>           path: diff.html
>           retention-days: 3
> ```
> 
> The `on PR` one adds a comment to the PR with the html content of the diff. This one I haven't tested yet since it involves this PR being accepted in the repo I commited it and then a next PR failing prettier test.
> 
> ```yaml
> name: Prettier code formatter (PR)
> 
> on:
>   pull_request:
>     branches:
>       - master
>       - main
>   workflow_dispatch:
> 
> jobs:
>   check:
>     # available images: https://github.com/actions/runner-images#available-images
>     runs-on: ubuntu-latest
>     steps:
>       - name: Checkout 🛎️
>         uses: actions/checkout@v4
>       - name: Setup Node.js ⚙️
>         uses: actions/setup-node@v4
>       - name: Install Prettier 💾
>         run: npm install --save-dev --save-exact prettier @shopify/prettier-plugin-liquid
>       - name: Prettier Check 🔎
>         id: prettier
>         run: npx prettier . --check
>       - name: Show diff 📝
>         # https://docs.github.com/en/actions/learn-github-actions/expressions#failure
>         if: ${{ failure() }}
>         run: |
>           npx prettier . --write
>           git diff -- . ':(exclude)package-lock.json' ':(exclude)package.json' > diff.txt
>           npm install -g diff2html-cli
>           diff2html -i file -s side -F diff.html -- diff.txt
>       - name: PR comment with html diff
>         # https://docs.github.com/en/actions/learn-github-actions/expressions#failure-with-conditions
>         if: ${{ failure() && steps.prettier.conclusion == 'failure' }}
>         uses: thollander/actions-comment-pull-request@v2
>         with:
>           filePath: diff.html
> ```

**Comment by mark-wiemer** [comment]
In short, the best workaround I've found is to run `prettier --write . && git diff` and go from there. Your favorite diff viewer will help too :)

**Comment by webark** [comment]
@mark-wiemer our of curiosity, what's the benefiting of needing to do a git diff over just using the prettier lint plugins?

https://prettier.io/docs/en/integrating-with-linters.html

**Comment by fisker** [comment]
I'm fine to add diff, but better to implement it as a "report", user can choose how the result to be reported.

**Comment by mwiemer-microsoft** [comment]
@webark I'm no prettier expert--if you find something that works and you like it, go for it. Personally I lint my code for everything except formatting, so I'm not the guy to ask. I found `git diff` to be the simplest way to get what I wanted, but it may not be best for everyone

(same person, now my work acct)

**Comment by george-gca** [comment]
If you need Prettier for CI cases, one option I just found out is [pr-prettier](https://github.com/abdulrehman-11/pr-prettier). Apparently it can add a commit to a PR with the prettified code.

**Comment by jacobrayschwartz** [comment]
Throwing my hat into the ring. Sometimes we are dealing with instances where the results in CI are different from the results we see locally. Having a diff available to report on makes understanding what is going on much faster. 

The different results might be coming from any number of issues, including resolving different versions of the library in CI from local development, true. But this simple output would help to cut down a lot of time resolving these types of issues. 

Also, knowing what issues prettier detects and understanding **why** they are issues helps developers move faster and format their code according to those practices by habit.

**Comment by jacobrayschwartz** [comment]
And adding my workaround, we use `yarn format` and `yarn format:fix` to run `prettier --check` and `prettier --write`. Just in case anyone finds this useful to get themselves unblocked, it ain't pretty, but gets the job done:

```sh
yarn format && echo 'format successful' || (yarn format:fix; git diff; echo 'yarn format failed, review diff above or run "yarn format:fix"'; return 1)
```

**Comment by yermulnik** [comment]
Would really appreciate `--diff` arg to make it obvious what exactly wasn't "pretty" in original file. Thanks.

**Comment by galangel** [comment]
+1 for diff

**Comment by BeyondEvil** [comment]
This is the definition of bike shedding 🤣 

A 5 year discussion to add functionality that most, if not all, other linters/formatters comes with. 😅

**Comment by trusktr** [comment]
> I personally like the output of `--check` very much.
> 
> _CI:_ "Forgot to run Prettier?"
> 
> _Me:_ "Oh yeah, I guess I did!" _Runs Prettier._
> 
> _CI:_ "All good!"

My case is:

_CI:_ "Forgot to run Prettier?"

_Me:_ "What the?" _Runs Prettier locally, no errors._

_CI:_ "Forgot to run Prettier?"

---

A diff would be helpful.

**Comment by lydell** [comment]
I added this to the first post in this thread to summarize:

> This is a very long thread. It starts out with me asking for use cases for this feature. People then provide use cases over and over. It is clear that this is a wanted feature. There is no need to add more comments about that, but do react with 👍 on this post. There is a pull request https://github.com/prettier/prettier/pull/12598 for this feature, but it is (in my opinion) a bit lacking: https://github.com/prettier/prettier/pull/12598#issuecomment-2338648470. I’m not a maintainer anymore, but it seems to me like a good PR for this would be merged.
